File: examples/python/common/event_logger.py
"""Common event logging utilities for projectors."""

from proto import domains_pb2 as domains


# ANSI color codes
BLUE = "\033[94m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
CYAN = "\033[96m"
MAGENTA = "\033[95m"
RED = "\033[91m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"


def domain_color(domain: str) -> str:
    """Get color for domain."""
    return BLUE if domain == "customer" else MAGENTA


def event_color(event_type: str) -> str:
    """Get color for event type."""
    if "Created" in event_type:
        return GREEN
    elif "Completed" in event_type:
        return CYAN
    elif "Cancelled" in event_type:
        return RED
    elif "Added" in event_type or "Applied" in event_type:
        return YELLOW
    return ""


def log_event(domain: str, root_id: str, sequence: int, type_url: str, data: bytes) -> None:
    """Log a single event with pretty formatting."""
    event_type = type_url.split(".")[-1] if type_url else "Unknown"

    # Header
    print()
    print(f"{BOLD}{'─' * 60}{RESET}")
    print(
        f"{BOLD}{domain_color(domain)}[{domain.upper()}]{RESET} "
        f"{DIM}seq:{sequence}{RESET}  "
        f"{CYAN}{root_id}...{RESET}"
    )
    print(f"{BOLD}{event_color(event_type)}{event_type}{RESET}")
    print(f"{'─' * 60}")

    # Parse and print event details
    _print_event_details(event_type, data)


def _print_event_details(event_type: str, data: bytes) -> None:
    """Parse and print event-specific details."""
    if event_type == "CustomerCreated":
        event = domains.CustomerCreated()
        event.ParseFromString(data)
        print(f"  {DIM}name:{RESET}    {event.name}")
        print(f"  {DIM}email:{RESET}   {event.email}")
        if event.HasField("created_at"):
            ts = event.created_at.ToDatetime()
            print(f"  {DIM}created:{RESET} {ts.isoformat()}")

    elif event_type == "LoyaltyPointsAdded":
        event = domains.LoyaltyPointsAdded()
        event.ParseFromString(data)
        print(f"  {DIM}points:{RESET}      +{event.points}")
        print(f"  {DIM}new_balance:{RESET} {event.new_balance}")
        print(f"  {DIM}reason:{RESET}      {event.reason}")

    elif event_type == "LoyaltyPointsRedeemed":
        event = domains.LoyaltyPointsRedeemed()
        event.ParseFromString(data)
        print(f"  {DIM}points:{RESET}      -{event.points}")
        print(f"  {DIM}new_balance:{RESET} {event.new_balance}")
        print(f"  {DIM}type:{RESET}        {event.redemption_type}")

    elif event_type == "TransactionCreated":
        event = domains.TransactionCreated()
        event.ParseFromString(data)
        print(f"  {DIM}customer:{RESET} {event.customer_id[:16]}...")
        print(f"  {DIM}items:{RESET}")
        for item in event.items:
            line_total = item.quantity * item.unit_price_cents
            print(
                f"    - {item.quantity}x {item.name} "
                f"@ ${item.unit_price_cents / 100:.2f} = ${line_total / 100:.2f}"
            )
        print(f"  {DIM}subtotal:{RESET} ${event.subtotal_cents / 100:.2f}")

    elif event_type == "DiscountApplied":
        event = domains.DiscountApplied()
        event.ParseFromString(data)
        print(f"  {DIM}type:{RESET}     {event.discount_type}")
        print(f"  {DIM}value:{RESET}    {event.value}")
        print(f"  {DIM}discount:{RESET} -${event.discount_cents / 100:.2f}")
        if event.coupon_code:
            print(f"  {DIM}coupon:{RESET}   {event.coupon_code}")

    elif event_type == "TransactionCompleted":
        event = domains.TransactionCompleted()
        event.ParseFromString(data)
        print(f"  {DIM}total:{RESET}    ${event.final_total_cents / 100:.2f}")
        print(f"  {DIM}payment:{RESET}  {event.payment_method}")
        print(f"  {DIM}loyalty:{RESET}  +{event.loyalty_points_earned} pts")

    elif event_type == "TransactionCancelled":
        event = domains.TransactionCancelled()
        event.ParseFromString(data)
        print(f"  {DIM}reason:{RESET} {event.reason}")

    else:
        print(f"  {DIM}(raw bytes: {len(data)} bytes){RESET}")


def project_events(event_book: dict) -> None:
    """Process all events in an event book and log them."""
    cover = event_book.get("cover", {})
    domain = cover.get("domain", "unknown")
    root_id = cover.get("root", {}).get("value", b"").hex()[:16]

    pages = event_book.get("pages", [])
    for page in pages:
        event = page.get("event", {})
        type_url = event.get("type_url", "")
        event_data = event.get("value", b"")
        sequence = page.get("sequence", {}).get("num", 0)

        log_event(domain, root_id, sequence, type_url, event_data)

File: examples/go/angzarr/event_logger.go
// Package angzarr provides shared event logging utilities for projectors.
package angzarr

import (
	"fmt"
	"strings"

	goproto "google.golang.org/protobuf/proto"

	"angzarr/proto/examples"
)

// ANSI color codes
const (
	Blue    = "\033[94m"
	Green   = "\033[92m"
	Yellow  = "\033[93m"
	Cyan    = "\033[96m"
	Magenta = "\033[95m"
	Red     = "\033[91m"
	Bold    = "\033[1m"
	Dim     = "\033[2m"
	Reset   = "\033[0m"
)

// DomainColor returns the color for a domain.
func DomainColor(domain string) string {
	if domain == "customer" {
		return Blue
	}
	return Magenta
}

// EventColor returns the color for an event type.
func EventColor(eventType string) string {
	switch {
	case strings.Contains(eventType, "Created"):
		return Green
	case strings.Contains(eventType, "Completed"):
		return Cyan
	case strings.Contains(eventType, "Cancelled"):
		return Red
	case strings.Contains(eventType, "Added"), strings.Contains(eventType, "Applied"):
		return Yellow
	default:
		return ""
	}
}

// LogEvent logs a single event with pretty formatting.
func LogEvent(domain, rootID string, sequence uint32, typeURL string, data []byte) {
	eventType := typeURL
	if idx := strings.LastIndex(typeURL, "."); idx >= 0 {
		eventType = typeURL[idx+1:]
	}

	// Header
	fmt.Println()
	fmt.Printf("%s%s%s\n", Bold, strings.Repeat("─", 60), Reset)
	fmt.Printf("%s%s[%s]%s %sseq:%d%s  %s%s...%s\n",
		Bold, DomainColor(domain), strings.ToUpper(domain), Reset,
		Dim, sequence, Reset,
		Cyan, rootID, Reset)
	fmt.Printf("%s%s%s%s\n", Bold, EventColor(eventType), eventType, Reset)
	fmt.Println(strings.Repeat("─", 60))

	// Event-specific details
	PrintEventDetails(eventType, data)
}

// PrintEventDetails parses and prints event-specific details.
func PrintEventDetails(eventType string, data []byte) {
	switch eventType {
	case "CustomerCreated":
		var event examples.CustomerCreated
		if err := goproto.Unmarshal(data, &event); err == nil {
			fmt.Printf("  %sname:%s    %s\n", Dim, Reset, event.Name)
			fmt.Printf("  %semail:%s   %s\n", Dim, Reset, event.Email)
			if event.CreatedAt != nil {
				fmt.Printf("  %screated:%s %s\n", Dim, Reset, event.CreatedAt.AsTime().Format("2006-01-02T15:04:05"))
			}
		}

	case "LoyaltyPointsAdded":
		var event examples.LoyaltyPointsAdded
		if err := goproto.Unmarshal(data, &event); err == nil {
			fmt.Printf("  %spoints:%s      +%d\n", Dim, Reset, event.Points)
			fmt.Printf("  %snew_balance:%s %d\n", Dim, Reset, event.NewBalance)
			fmt.Printf("  %sreason:%s      %s\n", Dim, Reset, event.Reason)
		}

	case "LoyaltyPointsRedeemed":
		var event examples.LoyaltyPointsRedeemed
		if err := goproto.Unmarshal(data, &event); err == nil {
			fmt.Printf("  %spoints:%s      -%d\n", Dim, Reset, event.Points)
			fmt.Printf("  %snew_balance:%s %d\n", Dim, Reset, event.NewBalance)
			fmt.Printf("  %stype:%s        %s\n", Dim, Reset, event.RedemptionType)
		}

	case "OrderCreated":
		var event examples.OrderCreated
		if err := goproto.Unmarshal(data, &event); err == nil {
			custID := event.CustomerId
			if len(custID) > 16 {
				custID = custID[:16]
			}
			fmt.Printf("  %scustomer:%s %s...\n", Dim, Reset, custID)
			fmt.Printf("  %sitems:%s\n", Dim, Reset)
			for _, item := range event.Items {
				lineTotal := item.Quantity * item.UnitPriceCents
				fmt.Printf("    - %dx %s @ $%.2f = $%.2f\n",
					item.Quantity, item.Name,
					float64(item.UnitPriceCents)/100,
					float64(lineTotal)/100)
			}
			fmt.Printf("  %ssubtotal:%s $%.2f\n", Dim, Reset, float64(event.SubtotalCents)/100)
		}

	case "LoyaltyDiscountApplied":
		var event examples.LoyaltyDiscountApplied
		if err := goproto.Unmarshal(data, &event); err == nil {
			fmt.Printf("  %spoints:%s  %d\n", Dim, Reset, event.PointsUsed)
			fmt.Printf("  %sdiscount:%s -$%.2f\n", Dim, Reset, float64(event.DiscountCents)/100)
		}

	case "OrderCompleted":
		var event examples.OrderCompleted
		if err := goproto.Unmarshal(data, &event); err == nil {
			fmt.Printf("  %stotal:%s    $%.2f\n", Dim, Reset, float64(event.FinalTotalCents)/100)
			fmt.Printf("  %spayment:%s  %s\n", Dim, Reset, event.PaymentMethod)
			fmt.Printf("  %sloyalty:%s  +%d pts\n", Dim, Reset, event.LoyaltyPointsEarned)
		}

	case "OrderCancelled":
		var event examples.OrderCancelled
		if err := goproto.Unmarshal(data, &event); err == nil {
			fmt.Printf("  %sreason:%s %s\n", Dim, Reset, event.Reason)
		}

	default:
		fmt.Printf("  %s(raw bytes: %d bytes)%s\n", Dim, len(data), Reset)
	}
}
